### PR TITLE
Add redirect persistence storage to auth compat layer

### PR DESCRIPTION
**Pull request by samhorlbeck**
b/166287735

### CODE REVIEW

File: packages-exp/auth-compat-exp/src/auth.ts [hunk 1]
```diff
@@ -46,10 +48,15 @@ export class Auth
...
     }
 
+    // Note this is slightly different behavior: in this case, the stored
+    // persistence is checked *first* rather than last. This is because we want
+    // the fallback (if no user is found) to be the stored persistence type
+    const storedPersistence = this.getPersistenceFromRedirect();
+    const persistences = storedPersistence ? [storedPersistence] : [];
+    persistences.push(impl.indexedDBLocalPersistence);
+
     // TODO(avolkovi): Implement proper persistence fallback
```
Comment (avolkovi) [inline packages-exp/auth-compat-exp/src/auth.ts:58]: do we have a bug for this?

File: packages-exp/auth-compat-exp/src/auth.ts [hunk 2]
```diff
@@ -297,6 +305,49 @@ export class Auth
...
   }
+
+  private savePersistenceForRedirect(): void {
+    const win = getSelfWindow();
+    const key = impl._persistenceKeyName(
+      PERSISTENCE_KEY,
+      this.auth.config.apiKey,
+      this.auth.name
+    );
+    if (win && win.sessionStorage) {
```
Comment (avolkovi) [inline packages-exp/auth-compat-exp/src/auth.ts:316]: win?.sessionStorage

File: packages-exp/auth-compat-exp/src/auth.ts [hunk 3]
```diff
@@ -297,6 +305,49 @@ export class Auth
...
+      case impl.browserLocalPersistence.type:
+        return impl.browserLocalPersistence;
+      default:
+        return null;
+    }
+  }
+}
+
+function getSelfWindow(): Window | null {
+  return typeof self !== 'undefined' ? self : null;
```
Comment (avolkovi) [inline packages-exp/auth-compat-exp/src/auth.ts:350]: typeof window !== 'undefined'

none of this will work in node

**Review by avolkovi — APPROVED** [review]

File: packages-exp/auth-compat-exp/src/auth.ts (hunk 1 above)
Comment (samhorlbeck) [reply packages-exp/auth-compat-exp/src/auth.ts:58]: b/170971681

**Pull request merged**